versioning/worker-versions.md

```diff
@@ -0,0 +1,174 @@
+
+## How is versioning accomplished with the `workerBuildId`?
+
+Every new worker ID will be assumed to be incompatible with the existing id(s)
+unless explicitly specified otherwise. New workers will not process *anything* 
+until explicitly enabled if they opt into build-id based versioning. Users who 
+wish to always have new workers begin processing new workflows can make an API 
+call immediately establishing the new worker as latest as part of their 
+deployment process (see below).
+
+In order to either (safely or not) migrate *open* workflows to a new version,
+users will explicitly indicate that an ID is compatible with some already-extant
+ID.
+
+Because this behavior is a breaking change, SDKs will need to add a
+`enableBuildIdBasedVersioning` boolean flag to their `WorkerOptions`.
+
+## Server-side APIs for `workerBuildId`-based versioning
+
+Importantly, we _already_ send the id (checksum, today) on every poll request.
+It's TBD if the worker will be changed to poll a unique queue per ID, or if
+that will happen on the frontend, or matching.
+
+These APIs could very well end up in a different service, but are shown
+as part of `WorkflowService` here for simplicity.
+```protobuf
+service WorkflowService {
+    rpc SetWorkerBuildIdOrdering (SetWorkerBuildIdOrderingRequest) returns (SetWorkerBuildIdOrderingResponse) {}
+    // This could / maybe should just be part of `DescribeTaskQueue`, but is broken out here to show easily.
+    rpc GetWorkerBuildIdOrdering (GetWorkerBuildIdOrderingRequest) returns (GetWorkerBuildIdOrderingResponse) {}
+}
+
+message SetWorkerBuildIdOrderingRequest {
+    string namespace = 1;
+    // Must be set, the task queue to apply changes to.
+    string task_queue = 2;
+    // The build id we are targeting.
+    string worker_build_id = 3;
+    // When set, indicates that the `workerBuildId` in this message is compatible
+    // with the one specified in this field. Because compatability should form
+    // a DAG, any build id can only be the "next compatible" version for one
+    // other ID at a time, and any setting which would create a cycle is invalid.
+    string is_compatible_with = 4;
+    // When set, establishes the specified `workerBuildId` as the latest
+    // for the queue. Those workers will begin processing new workflow executions.
+    bool latest = 5
+}
+message SetWorkerBuildIdOrderingResponse {}
+
+message GetWorkerBuildIdOrderingRequest {
+    string namespace = 1;
+    // Must be set, the task queue to interrogate about worker id ordering
+    string task_queue = 2;
+    // Limits how deep the returned DAG will go. 1 will return only the
+    // latest build id.
+    uint32 max_depth = 3;
+}
+message GetWorkerBuildIdOrderingResponse {
+    // The currently established latest worker build id
+    WorkerBuildIdNode current_latest = 1;
+    // Other current latest-compatible versions who are not the overall latest
+    repeated WorkerBuildIdNode compatible_leaves = 2;
```

Comment:
Can you clarify this a bit? Say I have the following worker build IDs for a task queue:

* `foo` - compatible with `old_foo1`, latest=true
* `old_foo1` - not compatible with anything
* `bar` - compatble with `old_bar1`
* `old_bar1` - not compatible with anything

How do I retrieve `bar`? Is that what shows up here?

Reply:
`bar` will be in `compatible_leaves` , and have a pointer back to `old_bar1`

You raise a good point though that if I send an ordering request for something that is not compatible with anything, and also not latest, that should be an error since it's just going to be some free-floating node.